gobblin-runtime/src/main/java/gobblin/runtime/SafeDatasetCommit.java

```diff
@@ -129,27 +150,62 @@ public Void call()
   }
 
   /**
-   * Synchronized version of {@link #commitDataset(JobState.DatasetState, DataPublisher)} used when publisher is not
+   * Synchronized version of {@link #commitDataset(Collection, DataPublisher)} used when publisher is not
    * thread safe.
    */
-  private synchronized void safeCommitDataset(JobState.DatasetState datasetState, DataPublisher publisher) {
-    commitDataset(datasetState, publisher);
+  private void safeCommitDataset(Collection<TaskState> taskStates, DataPublisher publisher) {
+    synchronized (GLOBAL_LOCK) {
```

Comment:
Publishers are not necessarily thread safe. In line 108 we check whether the publisher is thread-safe, if it is we don't use syncronization, but if it isn't we do.

Previously the synchronization was mistakenly done only within a single dataset, which is a mistake. This just ensures global synchronization when the publisher is not thread safe.